kafka/integration_test.go

```diff
@@ -0,0 +1,529 @@
+		t.Fatalf("Failed to create producer: %s\n", err)
+	}
+
+	t.Logf("Created Producer %v\n", p)
+
+	deliveryChan := make(chan Event)
+
+	for _, testmsg := range testmsgs {
+		m := &Message{
+			TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: testmsg.Partition},
+			Value:          testmsg.Value,
+			Key:            testmsg.Key}
+
+		// coverage test for message String() API
+		msgContent := m.String()
+		t.Logf("Producing message %s\n", msgContent)
+
+		err = p.Produce(m, deliveryChan)
+		if err != nil {
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
+				fmt.Fprintf(os.Stderr, "OK: Expected error: %s\n", testmsg.ExpectedError)
+				continue
+			} else {
+				t.Errorf("Encountered error: %s\n", err)
+				continue
+			}
+		}
+
+		e := <-deliveryChan
+		verifyEvent(e, testmsg, t)
+	}
+	p.Close()
+}
+
+// TestProducerChannelBasedAPI test channel-based API, requires broker
+func TestProducerChannelBasedAPI(t *testing.T) {
+	runProducerChannelBasedAPI(t)
+}
+
+// run the prodcuer using channel based API
+func runProducerChannelBasedAPI(t *testing.T) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+
+	kConfig := &ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"default.topic.config": ConfigMap{"produce.offset.report": true}}
+
+	p, err := NewProducer(kConfig)
+	if err != nil {
+		t.Fatalf("Failed to create producer: %s\n", err)
+	}
+
+	t.Logf("Created Producer %v\n", p)
+
+	var doneChan chan bool
+
+	for _, testmsg := range testmsgs {
+
+		doneChan = make(chan bool)
+
+		// delivery handler
+		go func() {
+			e := <-p.Events()
+			verifyEvent(e, testmsg, t)
+			close(doneChan)
+		}()
+
+		m := &Message{
+			TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: testmsg.Partition},
+			Value:          testmsg.Value,
+			Key:            testmsg.Key}
+		p.ProduceChannel() <- m
+
+		// wait for delivery report goroutine to finish
+		_ = <-doneChan
+
+	}
+
+	p.Close()
+
+}
+
+// verify event info
+func verifyEvent(e Event, testmsg testmsgType, t *testing.T) {
+	switch e.(type) {
+	case *Message:
+		m := e.(*Message)
+		if m.TopicPartition.Error != nil {
+			err := m.TopicPartition.Error
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
+				fmt.Fprintf(os.Stderr, "OK: Expected error: %s\n", testmsg.ExpectedError)
+			} else {
+				t.Errorf("Encountered unexepected error: %s\n", err)
+			}
+			return
+
+		}
+		// check partition
+		if !((testmsg.Partition == PartitionAny && m.TopicPartition.Partition >= 0) ||
+			(testmsg.Partition >= 0 && m.TopicPartition.Partition == testmsg.Partition)) {
+			t.Errorf("Partition check failed for test message %v. Got back {[%d] [%s] [%s]}\n",
+				testmsg, m.TopicPartition.Partition, m.Key, m.Value)
+			return
+
+		}
+
+		// check Key
+		if testmsg.Key == nil {
+			if !(m.Key == nil || len(m.Key) == 0) {
+				t.Errorf("Key verification failed. Expected %s, got %s\n", testmsg.Key, m.Key)
+				return
```

Comment:
If we remove the 'return's in this fail clauses we could get all errors reported for a message, which might help troubleshooting.